feat(proposal): enhanced worker auth proposal (#770)
go-vela/community: proposals/2023/03-15_worker-auth.md

While there may be other forms of authentication available for the server and worker, there already exists the infrastructure to use JWT tokens with claims on expiration, token type, etc.

4. Are there any related issues? Please provide them below if any exist.

<!-- Answer here -->

N/A

## Design

<!--
This section is intended to explain the solution design for the proposal.

NOTE: If there are no current plans for a solution, please leave this section blank.
-->

**Please describe your solution to the proposal. This includes, but is not limited to:**

* new/updated endpoints or url paths
* new/updated configuration variables (environment, flags, files, etc.)
* performance and user experience trade-offs
* security concerns or assumptions
* examples or (pseudo) code snippets

<!-- Answer here -->

A diagram will be added to this proposal soon to provide a visual aid for the design.

### Server Design

With the implementation of the token manager and build tokens, the ground work is already present to mint and validate worker auth tokens.

Some of the major changes will be:

- Add the ability for the `internal/token` package to mint and validate worker auth / register tokens

> - Add a `GET` registration endpoint, which will be platform admin exclusive. It will return a worker registration token.

Reviewer: This returns the token the admin will use to give the worker to use as its first token, right?

Each request will return a new token, right? If so, I think this should be a POST instead, as that better matches the HTTP verb's general usage. GETs should generally be cache-able by intermediate hosts.

Author: Sounds good, I agree

### Worker Design

The worker design is more tricky, but will look something like this:

1. On start up, the worker will enter a dead loop, where it will do nothing (no pulling from queue, checking in, etc) except serve the worker API. It will then be the job of the platform administrators to "register" the worker by hitting the worker's self registration endpoint.
2. Once the self-registration endpoint is hit, the worker will use the token gathered from the header of that incoming request and break its dead loop, thereby entering its check in loop and exec loop.
3-a. During the check in process, the worker will exchange its registration token for an auth token, which it will then refresh on the check-in interval. The success status of the check in will be a new struct field for the worker.
3-b. Based on the success status, the worker will proceed with its exec loop, checking before each pull whether the worker is truly verified with the server.
4. When a build is pulled off the queue, the worker will request a build token using its auth token.

> 5. In a scenario where a worker loses connection for a certain time period that renders the auth token expired, and admin is require to re-register the worker

Reviewer: ```suggestion
5. In a scenario where a worker loses connection for a certain time period that renders the auth token expired, and admin must re-register the worker.
```

Other considerations to further improve security:
- specification of IP address in the claims to allow restriction by IP
- stricter expiration logic

> In the case where the admins would prefer to continue using the symmetric token, all the above worker design logic will use the `VELA_SERVER_SECRET` token, which will not expire and not hinder executors' ability to pull from the queue.

Reviewer: Do you mean that the `VELA_SERVER_SECRET` can be used as a persistent worker auth token?

Author: Yes, the idea is that if it's provided (and there is a matching `VELA_SECRET` in the server env) then the server will honor that as a non-expiring auth method

### CLI Design

Since the registration token is so quick to expire, it is important to automate the job of onboarding workers for platform admins. This means that a new CLI command will be added that will use platform admin credentials to retrieve a registration token and hit the worker self-register endpoint. This will be vital for use in automation tools.

### Other things of note

- The server secret would be dropped from the container environment altogether if platform admins opt in to this enhancement.
- All auth tokens (check in + registration) have a strict expiration.
- Registration tokens can only be generated by a platform administrator

> - Registration tokens should expire very quickly

Reviewer: Any suggestions about how quickly? Would this be configurable?

Author: The default will be `1m` since the accompanying CLI changes will be quick enough to not have that be a problem. But yes it will be configurable.
